npiperelay.go

```diff
@@ -80,6 +83,110 @@ func dialPipe(p string, poll bool) (*overlappedFile, error) {
 	}
 }
 
+func dialPort(p int, _ bool) (*overlappedFile, error) {
+	if p < 0 || p > 65535 {
+		return nil, errors.New("invalid port value")
+	}
+
+	h, err := windows.Socket(windows.AF_INET, windows.SOCK_STREAM, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a SockaddrInet4 for connecting to
+	sa := &windows.SockaddrInet4{Addr: [4]byte{0x7F, 0x00, 0x00, 0x01}, Port: p}
+
+	// Bind to a randomly assigned local port
+	err = windows.Bind(h, &windows.SockaddrInet4{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap our socket up to be properly handled
+	conn := newOverlappedFile(h)
+
+	// Connect to the LibAssuan socket using overlapped ConnectEx operation
+	_, err = conn.asyncIo(func(h windows.Handle, _ *uint32, o *windows.Overlapped) error {
+		return windows.ConnectEx(h, sa, nil, 0, nil, o)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// LibAssaun file socket: Attempt to read contents of the target file and connect to a TCP port
+func dialAssuan(p string, poll bool) (*overlappedFile, error) {
+	pipeConn, err := dialPipe(p, poll)
+	if err != nil {
+		return nil, err
+	}
+
+	var port int
+	var nonce [16]byte
+
+	reader := bufio.NewReader(pipeConn)
+
+	// Read the target port number from the first line
+	tmp, _, err := reader.ReadLine()
+	if err == nil {
+		port, err = strconv.Atoi(string(tmp))
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the rest of the nonce from the file
+	n, err := reader.Read(nonce[:])
+	if err != nil {
+		return nil, err
+	}
+
+	if n != 16 {
+		err = fmt.Errorf("Read incorrect number of bytes for nonce. Expected 16, got %d (0x%X)", n, nonce)
+		return nil, err
+	}
+
+	if *verbose {
+		log.Printf("Port: %d, Nonce: %X", port, nonce)
+	}
+
+	err = pipeConn.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		// Try to connect to the libassaun TCP socket hosted on localhost
+		conn, err := dialPort(port, poll)
+
+		if poll && (err == windows.WSAETIMEDOUT || err == windows.WSAECONNREFUSED || err == windows.WSAENETUNREACH || err == windows.ERROR_CONNECTION_REFUSED) {
+			time.Sleep(cPOLL_TIMEOUT)
+			continue
+		}
+
+		if err != nil {
+			err = os.NewSyscallError("ConnectEx", err)
+			return nil, err
+		}
+
+		_, err = conn.Write(nonce[:])
+		if err != nil {
+			return nil, err
+		}
+
+		return conn, nil
+	}
+}
+
+func underlyingError(err error) error {
+	if serr, ok := err.(*os.SyscallError); ok {
+		return serr.Err
+	}
+	return err
+}
```

Comment:
I removed this function in my fork (https://github.com/albertony/npiperelay/commit/6ec21f6d7efbe654aa21d2618860984697981833), so it shouldn't be reintroduced.